File: src/http/identity/authentication/password_policy.rs
const MIN_PASSWORD_LENGTH: usize = 12;
const MAX_BCRYPT_PASSWORD_BYTES: usize = 71;
pub(super) const PASSWORD_TOO_LONG_MESSAGE: &str =
    "Password must be at most 71 UTF-8 bytes for bcrypt hashing";

pub fn validate_password_strength(password: &str) -> Result<(), &'static str> {
    if password.len() < MIN_PASSWORD_LENGTH {
        return Err("Password must be at least 12 characters long");
    }

    if password.len() > MAX_BCRYPT_PASSWORD_BYTES {
        return Err(PASSWORD_TOO_LONG_MESSAGE);
    }

    let has_lowercase = password.chars().any(char::is_lowercase);
    let has_uppercase = password.chars().any(char::is_uppercase);
    let has_digit = password.chars().any(|c| c.is_ascii_digit());
    let has_symbol = password.chars().any(|c| !c.is_alphanumeric());

    if !(has_lowercase && has_uppercase && has_digit && has_symbol) {
        return Err("Password must include lowercase, uppercase, numeric, and symbol characters");
    }

    Ok(())
}

File: web/src/features/auth/shared/model/passwordPolicy.ts
export const passwordPolicyChecks = [
  {
    key: "length",
    label: "At least 12 characters",
    test: (password: string) => password.length >= 12,
  },
  {
    key: "lower",
    label: "Lowercase letter",
    test: (password: string) => /[a-z]/.test(password),
  },
  {
    key: "upper",
    label: "Uppercase letter",
    test: (password: string) => /[A-Z]/.test(password),
  },
  {
    key: "number",
    label: "Number",
    test: (password: string) => /[0-9]/.test(password),
  },
  {
    key: "symbol",
    label: "Symbol",
    test: (password: string) => /[^A-Za-z0-9]/.test(password),
  },
  {
    key: "bcrypt",
    label: "At most 71 UTF-8 bytes",
    test: (password: string) => new TextEncoder().encode(password).length <= 71,
  },
];

export type PasswordPolicyCheck = (typeof passwordPolicyChecks)[number];
export type PasswordPolicyResult = PasswordPolicyCheck & { met: boolean };

export function evaluatePasswordPolicy(password: string) {
  return passwordPolicyChecks.map((check) => ({ ...check, met: check.test(password) }));
}
